Ensure AWS ORM initialization and improve session configuration handling

viiccwen opened the pull request:
## Summary

This PR introduces two related fixes to stabilize ORM initialization and session handling in AWS-related utilities.

## Changes

1. **Add `_ensure_db_session` in `eks_get_token`**

   * Introduced `_ensure_db_session()` to explicitly initialize the ORM when `engine` or `Session` are `None`.
   * Ensures that running `eks_get_token` as a standalone CLI properly initializes Airflow settings before creating the `EksHook`.
   * Added corresponding unit test `test_ensure_db_session_initializes_orm` in `test_eks_get_token.py` to validate this behavior.

2. **Refactor config initialization in `base_aws`**

   * Moved configuration creation logic inside the `get_session` method.
   * This ensures lazy evaluation and avoids potential issues when `Session` is not yet available at import time.
   * Improves reliability when invoking AWS hooks in contexts where Airflow ORM is not pre-initialized.

## Files Changed

* `airflow/providers/amazon/aws/utils/eks_get_token.py`
* `airflow/providers/amazon/aws/utils/tests/test_eks_get_token.py`
* `airflow/providers/amazon/aws/hooks/base_aws.py`

## Motivation

Previously, invoking `eks_get_token` directly from CLI could fail because Airflow’s ORM (`settings.engine`, `settings.Session`) was not yet initialized. This PR ensures ORM initialization is explicit and reliable.

Additionally, deferring configuration creation inside `get_session` removes the risk of creating invalid or stale sessions during import, aligning with lazy evaluation best practices.

## Testing

* Added new unit test for `_ensure_db_session` to validate ORM initialization when `engine` and `Session` are `None`.

## Related Issues

resolves #53578 

## Reproduced Step

```
# 1) initialization
# airflow db migrate
# airflow connections add val_aws_assume_test \
  --conn-uri 'aws://FAKE_KEY:FAKE_SECRET@/?region_name=eu-central-1'

# 2) reproduced
# export AWS_EC2_METADATA_DISABLED=true
# export PYTHON_OPERATORS_VIRTUAL_ENV_MODE=1
# python -m airflow.providers.amazon.aws.utils.eks_get_token \
  --cluster-name dummy --region-name eu-central-1
```

in `main` branch:

```
[2025-08-16T16:13:59.043+0000] {base_aws.py:621} WARNING - Unable to find AWS Connection ID 'val_aws_assume_test', switching to empty.
[2025-08-16T16:13:59.044+0000] {base_aws.py:197} INFO - No connection ID provided. Fallback on boto3 credential strategy (region_name='\x1b[1meu-central-1\x1b[22m'). See: https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html
[2025-08-16T16:13:59.048+0000] {base_aws.py:197} INFO - No connection ID provided. Fallback on boto3 credential strategy (region_name='\x1b[1meu-central-1\x1b[22m'). See: https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html
[2025-08-16T16:13:59.619+0000] {base_aws.py:621} WARNING - Unable to find AWS Connection ID 'aws_default', switching to empty.
[2025-08-16T16:13:59.619+0000] {base_aws.py:197} INFO - No connection ID provided. Fallback on boto3 credential strategy (region_name='\x1b[1meu-central-1\x1b[22m'). See: https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/opt/airflow/providers/amazon/src/airflow/providers/amazon/aws/utils/eks_get_token.py", line 64, in <module>
    main()
  File "/opt/airflow/providers/amazon/src/airflow/providers/amazon/aws/utils/eks_get_token.py", line 58, in main
    access_token = eks_hook.fetch_access_token_for_cluster(args.cluster_name)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/airflow/providers/amazon/src/airflow/providers/amazon/aws/hooks/eks.py", line 648, in fetch_access_token_for_cluster
    signed_url = signer.generate_presigned_url(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/botocore/signers.py", line 355, in generate_presigned_url
    self.sign(
  File "/usr/local/lib/python3.12/site-packages/botocore/signers.py", line 200, in sign
    auth.add_auth(request)
  File "/usr/local/lib/python3.12/site-packages/botocore/auth.py", line 421, in add_auth
    raise NoCredentialsError()
botocore.exceptions.NoCredentialsError: Unable to locate credentials
```

in `fix-aws-orm-initialization`:
```
[2025-08-17T05:17:18.561+0000] {connection_wrapper.py:334} INFO - AWS Connection (conn_id='val_aws_assume_test', conn_type='aws') credentials retrieved from login and password.
[2025-08-17T05:17:19.043+0000] {base_aws.py:621} WARNING - Unable to find AWS Connection ID 'aws_default', switching to empty.
[2025-08-17T05:17:19.043+0000] {base_aws.py:197} INFO - No connection ID provided. Fallback on boto3 credential strategy (region_name='\x1b[1meu-central-1\x1b[22m'). See: https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html
expirationTimestamp: 2025-08-17T05:31:19Z, token: k8s-aws-v1.aHR0cHM6Ly9zdHMuZXUtY2VudHJhbC0xLmFtYXpvbmF3cy5jb20vP0FjdGlvbj1HZXRDYWxsZXJJZGVudGl0eSZWZXJzaW9uPTIwMTEtMDYtMTUmWC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1GQUtFX0tFWSUyRjIwMjUwODE3JTJGZXUtY2VudHJhbC0xJTJGc3RzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MTdUMDUxNzE5WiZYLUFtei1FeHBpcmVzPTYwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCUzQngtazhzLWF3cy1pZCZYLUFtei1TaWduYXR1cmU9NjI2NmNhMzRjOTAyMTE4ZjdhOTA5YWU1MGI3NzkzMzg4ZDBkNGZiYmZlYzhmZjkwN2NiMzJkODE5NWMzZjU0Yg
```

viiccwen committed:
fix(aws): add _ensure_db_session function to initialize orm

- Add `_ensure_db_session` to initialize orm when engine or session are None
- Add `test_ensure_db_session_initializes_orm` to test ORM initialization logic

viiccwen committed:
fix(aws): move config initialization inside get_session method

Move config creation inside get_session for better lazy evaluation to avoid session issue

prdai commented on providers/amazon/src/airflow/providers/amazon/aws/utils/eks_get_token.py:
```diff
@@ -51,9 +53,16 @@ def get_parser():
 
+def _ensure_db_session():
+    if not getattr(settings, "engine", None) or not getattr(settings, "Session", None):
+        configure_orm()
```
a few small suggestions: maybe we could rename `_ensure_db_session` to `_ensure_orm_configured`? i think it is more fitting for what the function is specifically doing, as well as a small doc string with what is specifically happening within this function would be nice

jason810496 replied:
No strong opinion for the naming, both work with me.
+1 for having the docstring.

viiccwen replied:
big thx for correcting! 🤩

jason810496 commented on providers/amazon/src/airflow/providers/amazon/aws/hooks/base_aws.py:
```diff
@@ -688,9 +688,10 @@ def account_id(self) -> str:
-            conn=self.conn_config, region_name=region_name, config=self.config
-        ).create_session(deferrable=deferrable)
+        config = self._config or botocore.config.Config()
+        return SessionFactory(conn=self.conn_config, region_name=region_name, config=config).create_session(
```
May I ask why do we need this change?

Since `AwsGenericHook.get_session` will get from `conn_config.botocore_config or botocore.config.Config()
`.
https://github.com/apache/airflow/blob/6da99f245634029d4c62a0b1d66f053b2d23637e/providers/amazon/src/airflow/providers/amazon/aws/hooks/base_aws.py#L678-L681

Then the `AwsGenericHook.conn_config` property will return `AwsConnectionWrapper`.

https://github.com/apache/airflow/blob/6da99f245634029d4c62a0b1d66f053b2d23637e/providers/amazon/src/airflow/providers/amazon/aws/hooks/base_aws.py#L617-L643

The `__post_init__` method of `AwsConnectionWrapper` will ensure initializing the `AwsConnectionWrapper.botocore_config` property.

https://github.com/apache/airflow/blob/6da99f245634029d4c62a0b1d66f053b2d23637e/providers/amazon/src/airflow/providers/amazon/aws/utils/connection_wrapper.py#L244-L249

IMO, the `AwsGenericHook.config` will not be `None` and will be initialized properly.

viiccwen replied:
Oh...My mistake. I missed that config already ensures a valid config. I’ve removed the redundant fallback. Thanks for pointing it out! 🙏

jason810496 commented on providers/amazon/src/airflow/providers/amazon/aws/utils/eks_get_token.py:
```diff
@@ -51,9 +53,16 @@ def get_parser():
 
 
+def _ensure_db_session():
+    if not getattr(settings, "engine", None) or not getattr(settings, "Session", None):
```
Having check on `engine` would be enough.

FYI:
https://github.com/apache/airflow/blob/6da99f245634029d4c62a0b1d66f053b2d23637e/airflow-core/src/airflow/settings.py#L420-L421

viiccwen replied:
thx for pointing out!

jason810496 reviewed (commented):
Nice! Thanks for the fix.

viiccwen committed:
refactor(eks_get_token): rename, simplify function and add docstring

- Rename _ensure_db_session to _ensure_orm_configured
- Simplify ORM check (remove session part)
- add docstring with inline comment
- Update corresponding test function name and calls

potiuk reviewed (changes requested):
I do not understand: where eks_get_token is going to be used an in what circumstances?

Airflow Providers in Airflow 3 are supposed to **not** use Airflow ORM at all. So it sounds strange that there is a tool in provider that expects to initialize it.

What's the purpose and scenario of running this tool ?

jason810496 reviewed (commented):
>  unless of course EKS cluster has the credentials which is obviously wrong according to the Airflow security model.

I see and agree with your point.  
I overlooked the key issue that _the EKS cluster would need to have database credentials in order to initialize the ORM_. Thank you for highlighting this.

> I think though there is a way better solution - that is also "proper" the credentials of AWS connection should be uploaded to the EKS cluster (temporarily) as secret and the PODs run on EKS should use that secret (and the secret should be deleted after the pod completes. You cannot reallly get the EKS cluster that is used there to communicate with TaskSDK because there will normally be no firewalls opened for the cluster worker to communicate with API server (not mentioning that similarly if you want to run ORM/DB access you will have to have the capability of the EKS cluster to communicate back with Airflow DB.


Yes, setting the token with the credentials from the AWS connection is a much more appropriate solution in this case, and it also aligns with the behavior of other providers.

So, @viiccwen, it would be helpful to refactor the `get_eks_token` logic to use the Airflow Connection `extra` field (perhaps named `eks_sa_token`). Thank you!

The implementation could follow the pattern used in `athena_sql` for retrieving the `driver` config from the `extra` field:  
https://github.com/apache/airflow/blob/9b46d76ee4fffdbefa7fa14aba5b7d8a8ec15f3b/providers/amazon/src/airflow/providers/amazon/aws/hooks/athena_sql.py#L146-L151

Additionally, we should document this breaking change for `EksHook`, and provide guidance for users on how to set up credentials using Airflow Connection. For example, users should create a Service Account in EKS first, then set `<SERVICE-ACCOUNT-TOKEN>` in the Airflow Connection as `eks_sa_token` in the `extra` field.

cc @o-nikolas @vincbeck
Do you have any comments or feedback on this?
